### ISSUE TITLE
How to use Stimulus Reflex Lifecycle Methods as per links in description using view_component_reflex?

### ISSUE THREAD

**Issue by omairrazam** [joshleblanc/view_component_reflex#65]
https://docs.stimulusreflex.com/rtfm/lifecycle#server-side-reflex-callbacks
https://docs.stimulusreflex.com/rtfm/lifecycle#custom-life-cycle-methods

Also view_component_reflex doesn't use sidecar stimulus controller but points to controller inside javascripts folder. How to manage that?

**Comment by joshleblanc** [comment]
You should be able to use the server side callbacks as if the component is a reflex. 

The location of the controller shouldn't matter for the frontend. It just assumes that the component name matches the name of the controller. 

Are you running into any problems in particular?

**Comment by omairrazam** [comment]
@joshleblanc Thanks for looking at my issue.

I tried to consider Component as a reflex and used standard way to call hook, but it didn't work. I will try again. if you have any working example that would be helpful.

secondly my controller is called as `data-controller="example-component--example-component"` if it is located inside my component but I see this is what is applied to my html from view_component_reflex `data-controller="example"` which means it is pointing to controller outside of component directory at a common place.

Really appreciate your help.

**Comment by joshleblanc** [comment]
@omairrazam I gave it a whirl on my end, and a component that looked like 

```ruby
# app/components/example_component/example_component.rb
class ExampleComponent::ExampleComponent < ViewComponentReflex::Component

end
```

Output the data controller `example-component--example`. What does your file structure look like?

**Comment by omairrazam** [comment]
<img width="1635" alt="Screenshot 2021-09-10 at 1 08 20 AM" src="https://user-images.githubusercontent.com/10624816/132755430-7d5ec290-bd58-46d2-b5d3-9b7b5a4d432d.png">

@joshleblanc really appreciate your help. Please see attached image.
Thanks.

**Comment by omairrazam** [comment]
@joshleblanc i generated my component using view_component gem's command and then manually changed parent class of component to what your gem suggests.

`rails generate component Example title --sidecar`

**Comment by joshleblanc** [comment]
Right now, there's no good way to generate the controller from the component, since it's not actually in the sidecar folder itself.

But you can use `component_controller data: { controller: "example-component--example-component" key: key }` to override view component reflex's defaults.

**Comment by omairrazam** [comment]
@joshleblanc overriding did work to pick my controller but my reflex stopped working, here is some clue from console. please see snapshot.

<img width="1267" alt="Screenshot 2021-09-18 at 5 46 50 AM" src="https://user-images.githubusercontent.com/10624816/133866367-b8150921-dc8f-42d9-ab85-1bf03bb4fdb2.png">

**Comment by omairrazam** [comment]
@joshleblanc ok by reading the code of gem, I am able to spot what to override and it worked.
I just override this class method

`def self.stimulus_controller
    'example-component--example-component'
  end`

**Comment by omairrazam** [comment]
The only issue now I need help with is to utilize before and after reflex hooks. I tried to use it as flex as mentioned in this doc.
https://docs.stimulusreflex.com/rtfm/lifecycle

But it didn't work.

@joshleblanc

**Comment by joshleblanc** [comment]
This is good stuff. Thanks for looking into this.

**Comment by joshleblanc** [comment]
@omairrazam So I tested this in the expo, and this worked as expected:

```ruby
class CounterComponent < ViewComponentReflex::Component
  after_reflex do
    p "Done" # This prints fine
  end
  
  def initialize
    @count = 0
  end

  def increment
    @count += 1
  end
end
```

Tried it using method references as well, which worked.

I did find a problem, though: Changing callbacks at runtime would not register. You would need to restart the server entirely. 
I imagine this is the problem you were running into.

I've just pushed 3.1.14.pre5 which should re-wire the callbacks on request in development. 
I've also updated the docs to include your fix for overriding the `self.stimulus_controller` method. 

Let me know if you run into anything else!
